web3/providers/rpc.py

```diff
@@ -57,6 +74,94 @@ def make_request(self, method, params):
         return response_body
 
 
+class KeepAliveRPCProvider(BaseProvider):
+    """RPC-provider that handles HTTP keep-alive connection correctly.
+
+    HTTP client and underlying TCP/IP network connection is recycled across requests.
+    :class:`HTTPClient` connection pooling is used for connections.
+
+    Preferably create only one instance of KeepAliveProvider per process,
+    though the class has internal in-process cache for clients.
+    """
+
+    #: In-process client cache keyed by host:port -> HTTPClient
+    clients = {}
```

Comment:
I *think* it would clean things up by moving this to a global variable at the module level which would allow extracting the `get_or_create_client` logic entirely from this class.  That would allow writing a very basic test for it as well without requiring much overhead.  Just a thought.  Not sure it's the right design decision.